rust/lance/src/dataset.rs

```diff
@@ -439,21 +441,53 @@ impl Dataset {
         batches: impl RecordBatchReader + Send + 'static,
         dest: impl Into<WriteDestination<'_>>,
         params: Option<WriteParams>,
+    ) -> Result<Self> {
+        let (batches, schema) = peek_reader_schema(Box::new(batches)).await?;
+        let stream = reader_to_stream(batches);
+        Self::write_stream(stream, schema, dest, params).await
+    }
+
+    /// Write to or Create a [Dataset] with a stream of [RecordBatch]s.
+    ///
+    /// `dest` can be a `&str`, `object_store::path::Path` or `Arc<Dataset>`.
+    ///
+    /// Returns the newly created [`Dataset`].
+    /// Or Returns [Error] if the dataset already exists.
+    ///
+    pub async fn write_stream(
+        stream: SendableRecordBatchStream,
+        schema: Schema,
+        dest: impl Into<WriteDestination<'_>>,
+        params: Option<WriteParams>,
     ) -> Result<Self> {
         let mut builder = InsertBuilder::new(dest);
         if let Some(params) = &params {
             builder = builder.with_params(params);
         }
-        builder.execute_stream(batches).await
+        builder.execute_stream(stream, schema).await
     }
 
-    /// Append to existing [Dataset] with a stream of [RecordBatch]s
+    /// Append to existing [Dataset] with an iterator of [RecordBatch]s
     ///
     /// Returns void result or Returns [Error]
     pub async fn append(
         &mut self,
         batches: impl RecordBatchReader + Send + 'static,
         params: Option<WriteParams>,
+    ) -> Result<()> {
+        let (batches, schema) = peek_reader_schema(Box::new(batches)).await?;
+        let stream = reader_to_stream(batches);
+        self.append_stream(stream, schema, params).await
+    }
+
+    /// Append to existing [Dataset] with a stream of [RecordBatch]s
+    ///
+    /// Returns void result or Returns [Error]
+    pub async fn append_stream(
+        &mut self,
+        stream: SendableRecordBatchStream,
+        schema: Schema,
+        params: Option<WriteParams>,
```

Comment:
I'd also rather not add these methods. If we keep adding these we'll end up with too many APIs. The `InsertBuilder` is public and I'd rather we use that. If we want, we could add `write_builder(&self) -> WriteBuilder;` and I'd be fine with that.

Reply:
Implemented by abstracting the (stream, schema) into `StreamingWriteSource`.